### ISSUE TITLE
Xpra 4.2/Debian: Deadlock, no connection possible

### ISSUE THREAD

**Issue by shururuun** [Xpra-org/xpra#3174]
I am debugging a strange phenomenon with xpra_4.2-r0-1_amd64.deb on Debian/Buster: It looks like Xpra starts, but deadlocks immediately, and the session can't be attached, neither from SSH nor from the HTML client. Killing it hard leaves the Xpra process as a zombie, seemingly depending on systemd, which for whatever reason doesn't reap that child.

Server start:
```xpra start :23 --systemd-run=no  --start-child=xterm --exit-with-children=yes --debug=all```

As far as I can interpret the debugging output, it looks like the server thinks the socket is already closed:

```
2021-06-21 10:59:26,220 org.xpra.Client(ClientConnectionMuxer(1 : Protocol(unix-domain socket:/run/user/1000/xpra/mimiru-23)))
2021-06-21 10:59:26,220 dbus_exception_wrap() <function DBUS_Mixin.init_state.<locals>.make_dbus_server at 0x7fca4b80c9d0>()=DBUS_Source(org.xpra.Server:/org/xpra/Client1)
2021-06-21 10:59:26,220 VideoHelper.init()
2021-06-21 10:59:26,221 peek: elapsed=25, timeout=100
2021-06-21 10:59:26,221 <built-in method recv of socket object at 0x7fcab56de580>(8192, MSG_PEEK)
2021-06-21 10:59:26,223 peek: elapsed=25, timeout=100
2021-06-21 10:59:26,223 <built-in method recv of socket object at 0x7fcab56de820>(8192, MSG_PEEK)
2021-06-21 10:59:26,246 peek: elapsed=50, timeout=100
2021-06-21 10:59:26,246 <built-in method recv of socket object at 0x7fcab56de580>(8192, MSG_PEEK)
2021-06-21 10:59:26,248 peek: elapsed=50, timeout=100
2021-06-21 10:59:26,248 <built-in method recv of socket object at 0x7fcab56de820>(8192, MSG_PEEK)
2021-06-21 10:59:26,271 peek: elapsed=75, timeout=100
2021-06-21 10:59:26,271 <built-in method recv of socket object at 0x7fcab56de580>(8192, MSG_PEEK)
2021-06-21 10:59:26,273 peek: elapsed=76, timeout=100
2021-06-21 10:59:26,273 <built-in method recv of socket object at 0x7fcab56de820>(8192, MSG_PEEK)
2021-06-21 10:59:26,296 peek: elapsed=101, timeout=100
2021-06-21 10:59:26,296 socket unix-domain socket:/run/user/1000/xpra/mimiru-23 peek: got 0 bytes
2021-06-21 10:59:26,296 socket peek=b''
2021-06-21 10:59:26,296 socket peek hex=
2021-06-21 10:59:26,297 socket peek line1=b''
2021-06-21 10:59:26,297 guess_packet_type(..)=None
2021-06-21 10:59:26,297 unix-domain connection already closed
2021-06-21 10:59:26,298 untilConcludes(<bound method Connection.is_active of unix-domain socket:/run/user/1000/xpra/mimiru-23>, <bound method Connection.can_retry of unix-domain socket:/run/user/1000/xpra/mimiru-23>, <built-in method recv of socket object at 0x7fcab56de700>, (65536,), {}) timed out, retry=socket.timeout
2021-06-21 10:59:26,298 peek: elapsed=101, timeout=100
2021-06-21 10:59:26,298 socket unix-domain socket:/home/fnord/.xpra/mimiru-23 peek: got 0 bytes
2021-06-21 10:59:26,298 socket peek=b''
2021-06-21 10:59:26,299 socket peek hex=
2021-06-21 10:59:26,299 socket peek line1=b''
2021-06-21 10:59:26,299 guess_packet_type(..)=None
2021-06-21 10:59:26,299 unix-domain connection already closed
```

If you shut down the Xpra server immediately after start with "xpra stop :23", then it times out as well.

The beta release 4.3-r29319-1 exhibits the same problem.

**Comment by shururuun** [comment]
After another round of debugging: This happens even with the command "xpra encoding":

stat("/dev/dri/card0", {st_mode=S_IFCHR|0660, st_rdev=makedev(0xe2, 0), ...}) = 0
stat("/dev/dri/renderD128", {st_mode=S_IFCHR|0660, st_rdev=makedev(0xe2, 0x80), ...}) = 0
openat(AT_FDCWD, "/usr/share/drirc.d", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 5
openat(AT_FDCWD, "/usr/share/drirc.d/00-mesa-defaults.conf", O_RDONLY) = 5
openat(AT_FDCWD, "/etc/drirc", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/user/.drirc", O_RDONLY) = -1 ENOENT (No such file or directory)
stat("/sys/dev/char/226:128/device/drm", {st_mode=S_IFDIR|0755, st_size=0, ...}) = 0
amdgpu_device_initialize: AMDGPU_INFO_ACCEL_WORKING = 0
amdgpu: amdgpu_device_initialize failed.

It deadlocks there.

It looks like trying to use the AMD GPU VAAPI driver deadlocks the GPU somehow, even if it working using the x264 VAAPI encoder in ffmpeg. This is probably somehow a GPU driver problem:

[525409.635239] [drm:amdgpu_job_timedout [amdgpu]] *ERROR* ring vcn_enc0 timeout, signaled seq=18, emitted seq=19
[525409.635592] [drm:amdgpu_job_timedout [amdgpu]] *ERROR* Process information: process  pid 0 thread  pid 0
[525409.635925] amdgpu 0000:0a:00.0: amdgpu: GPU reset begin!

Setting XPRA_VAAPI=0 avoids the problem.

**Comment by totaam** [comment]
Not the first time that hardware access causes hangs.
Perhaps we should make vaapi default to off.

**Comment by totaam** [comment]
I didn't want to disable vaapi everywhere by default so the best I could come up with was to use the kernel version and require something recent enough (5.0 or later) and hope that the drivers that come with it work well enough.

The end result is that VAAPI will be enabled by default on Ubuntu Focal (kernel 5.4.0) and later, Debian Bullseye (5.10.0) and later, but not on Debian Buster (4.19.0) or Ubuntu Bionic (4.15).
Hopefully that's good enough for now.

The proper solution to GPU driver crashes is #3191.

**Comment by totaam** [comment]
FYI: 4.2.1 has been released and vaapi is now disabled by default: fb52a4b8e58eaa27df509e2286d16efcd0d93ae6
